File: Session15/213. House Robber II.java
class Solution {
    public int rob(int[] nums) {
        int n=nums.length;
        if(n==1)
        {
            return nums[0];
        }
        if(n==0)
        {
            return 0;
        }
        // we are passing index
        return Math.max(rob_linear(0,nums,n-1),rob_linear(1,nums,n-1));
    }
    
     public int rob_linear(int start,int nums[],int n) {
        int dp[]=new int[n+2];
        dp[0]=0;
        dp[1]=0;
        
         // case I - Rob 0th house - start =0 end=n-2
         // case II - Do not Rob 0th House - start =1 end=n-1
        for(int i=2;i<n+2;i++)
        {
            dp[i]=Math.max(nums[start+i-2]+dp[i-2], dp[i-1]);
        }
        
        return dp[n+1];
    }
}

File: Session15/213. House Robber II.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        if len(nums) == 0 or nums is None:
            return 0

        if len(nums) == 1:
            return nums[0]

        return max(self.rob_simple(nums[:-1]), self.rob_simple(nums[1:]))

    def rob_simple(self, A: List[int]) -> int:
        if len(A) == 1: return A[0]
        dp = (len(A)+2)*[0]
        dp[0]=0
        dp[1] =0
        for i in range(2, len(A)+2):
            dp[i] = max(dp[i-1], A[i-2] + dp[i-2])
        return dp[-1]